webrtc/README.md

```diff
@@ -0,0 +1,445 @@
+   and _password_ in combination with the IP and port of _A_ can be used by _B_
+   to identify the connection, i.e. demultiplex incoming UDP datagrams per
+   incoming connection. _B_ uses the same random string for the username and
+   password in the STUN message from _B_ to _A_.
+
+   Note that this step requires _B_ to allocate memory for each incoming STUN
+   message from _A_. This could be leveraged for a DOS attack where _A_ is
+   sending many STUN messages with different ufrags using different UDP source
+   ports, forcing _B_ to allocate a new peer connection for each. _B_ SHOULD
+   have a rate limiting mechanism in place as a defense measure. See also
+   https://datatracker.ietf.org/doc/html/rfc5389#section-16.1.2.
+
+7. _A_ and _B_ execute the DTLS handshake as part of the standard WebRTC
+   connection establishment.
+
+   At this point _B_ does not know the TLS certificate fingerprint of _A_. Thus
+   _B_ can not verify _A_'s TLS certificate fingerprint during the DTLS
+   handshake. Instead _B_ needs to _disable certificate fingerprint
+   verification_ (see e.g. [Pion's `disableCertificateFingerprintVerification`
+   option](https://github.com/pion/webrtc/blob/360b0f1745c7244850ed638f423cda716a81cedf/settingengine.go#L62)).
+
+   On success of the DTLS handshake the connection provides confidentiality and
+   integrity but not authenticity. The latter is guaranteed through the
+   succeeding Noise handshake. See [Connection Security
+   section](#connection-security).
+
+8. Messages on the established `RTCDataChannel` are framed using the message
+   framing mechanism described in [Multiplexing](#multiplexing).
+
+9. The remote is authenticated via an additional Noise handshake. See
+   [Connection Security section](#connection-security).
+
+WebRTC can run both on UDP and TCP. libp2p WebRTC implementations MUST support
+UDP and MAY support TCP.
+
+## Multiplexing
+
+The WebRTC browser APIs do not support half-closing of streams nor resets of the
+sending part of streams.
+[`RTCDataChannel.close()`](https://developer.mozilla.org/en-US/docs/Web/API/RTCDataChannel/close)
+flushes the remaining messages and closes the local write and read side. After
+calling `RTCDataChannel.close()` one can no longer read from the channel. This
+lack of functionality is problematic, given that libp2p protocols running on top
+of transport protocols, like WebRTC, expect to be able to half-close or reset a
+stream. See [Connection Establishment in
+libp2p](https://github.com/libp2p/specs/blob/master/connections/README.md#definitions).
+
+To support half-closing and resets of streams, libp2p WebRTC uses message
+framing. Messages on a `RTCDataChannel` are embedded into the Protobuf message
+below and sent on the `RTCDataChannel` prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec].
```

Comment:
Random nit: the [LEB128](https://en.wikipedia.org/wiki/LEB128) format is widespread throughout the IT industry, I don't think we need to tell people to specifically go see in the multiformats specification.